Change in shooter angles while moving util

Panda8188 opened the pull request:
* Calculate change in "azimuth" angle based on tangential velocity
* Calculate change in "elevation" angle based on radial velocity

For shooter
*TODO: Implement util in shooter code given current tangential, radial velocity, distance from speaker, and current azimuth angle (needed unfortunately). simple tan logic.


google doc is being updated as of this post and will have a detailed explanation regarding all variables and related silly logic

The name "delta angle speed calc util" may be silly goofy, but can be understood as such:
Delta = change in
Angle = angle
(based on)
speed = speed/velocity
calc = calculation
util = utility

So, one can see: The change in angle based on speed is calculated in this utility

itsmesp8343 committed:
Created Swerve Follow Wrapper (#26)

(not rapper!)
copy from last year with changes to account for updated path stuff

balghane committed:
Fill in placeholder values (#24)

* Fill in placeholder values

* Reviewer comments

* Formatting fixes

---------

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

voidSevenSevenSix committed:
CRT Util (#12)

* Implemented CRT Util, added a unit test for it. I will add comments on the actual math later.

* Added some comments

* Math comments + one comment correction

* Basel Fixes I (not all fixes have been done yet)

* Fix build fail error

* Basel Fixes II (still not quite done, but class is ending)

* Basel Fixes III

* Basel Fixes IV - did the 10 comments github hid from me

* crt done, uncommenting unit tests

* typo

* Formatting fixes

---------

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

balghane committed:
Make sure elevator and shooter don't intersect (#30)

* Make sure elevator and shooter don't intersect

* Formatting fixes

* fix

* Formatting fixes

* reviewer comments

---------

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

tails618 committed:
Add intake limelight (#27)

* Add IntakeLimelight subsystem

Copied from 2023, some changes, not done yet.

* feat: Add note detection to IntakeLimelight

This commit adds the ability for the IntakeLimelight subsystem to detect notes. The `pipeline` enum now includes a new option for note detection. The `setPipeline()` method has been updated to set the pipeline correctly for note detection. A new inner class called `NoteDetection` has been added, which represents the detected note's latency, distance from the camera, and CCW angle. The `getConePos()` method has been renamed to `getNotePos()`, and it now returns an Optional<NoteDetection> instead of Optional<ConeDetection>. The `getAngleToConeDeg()` method has also been renamed to `getAngleToNoteDeg()`.

* more stuff from last year for apriltags

* One change I made during math class

* Address PR comments

* More changes

* Refactor IntakeLimelight code and add calibration constants

- Refactored the import statements to remove unused imports and added necessary imports for IntakeCal.
- Added a new import statement for LimelightHelpers.LimelightTarget_Detector.
- Updated the pitchAngleDegrees parameter description in the constructor of IntakeLimelight class.
- Removed unnecessary comments and empty lines.
- Added a new constant LIMELIGHT_YAW in IntakeLimelightCal class.
- Added two new constants INTAKE_LIMELIGHT_PITCH_DEGREES and INTAKE_LIMELIGHT_HEIGHT_METERS in IntakeLimelightConstants class.

* i forgor 💀 the note 🎵 has thickness 📏

* More changes

* Minor fix

* Formatting fixes

---------

Co-authored-by: James Hubbard <cobbsaladluv@gmail.com>
Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

tails618 committed:
Added shooter LL (identical to intake LL, just renamed "intake" stuff)

allison324 committed:
Light Class (#28)

* Light Class

We made Light class and mainly copied everything from last year's code(same functions and everything) - the only things that we changed are the colors and the enums. Also added a const int for the canid in RobotMap.java.

* Light class

minor changes, add lights constants

* Lights-class

Minor changes based on Ava's comments.

* LightsClassUpdates

I just went through all of the comments written by Ava and Asher the appropriate changes.

* Light class, added NO_TAG enum

* Formatting fixes

* delete drivetopoint

---------

Co-authored-by: ruben-hayrapetyan <ruben.hayrapetyan28@gmail.com>
Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

itsmesp8343 committed:
Set Brownout Voltage (#34)

* Set Brownout Voltage

set brownout voltage to 5

* Update

made var in constants

---------

Co-authored-by: allison324 <124639587+allison324@users.noreply.github.com>

SaanikaGW committed:
update Amp Shoot Command (#35)

* update Amp Shoot Command

Making it so amp shoot only waits for intake to be clear of the conveyor intead of waiting for the intake to be fully stowed

* Update AmpPrepScore.java

* PR comment

---------

Co-authored-by: Allison Li <ali@ucls.uchicago.edu>

avakimc committed:
Added Two Note Auto and Some Components (ScoreThisNote, DriveDistance) (#25)

* really basic outline of two note auto and a single component of ScoreThisGamePiece

added more stuff to basic two note auto

- added the command to drive to the second note, then the command to turn back around towards the speaker

Added Two Note Auto and Components

- DriveDistance.java which is a sequential command group that drives in a direction at a given speed for a certain distance
- added getMatchState() method in DriveSubsystem so that we can access the color in other classes

Edits and Added OneFive Auto

revisions

revisions 2/17

- pass in matchState as an instance variable
- added timeout to IntakeSequence
- make robot intake while drive (with a wait before drive)

* Formatting fixes

* Formatting fixes

---------

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

balghane committed:
Fixes from the 02/16 meetings (#32)

* Add second left motor to shooter

* Rename intake

* set can ids

* ctre status frame changes

* add subsystems to robot container

* Update elevator current limit, fix intake configurator

* Formatting fixes

* rename driver controller id

* fix typo

* Formatting fixes

---------

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

balghane committed:
Add burn flash sparks to Intake, Shooter, and Conveyor (#33)

avakimc committed:
Switched CanSparksMax to Coast when Disabled (#31)

* Switched CanSparksMax to Coast when Disabled

- made all CanSparksMax instance variables public
- made instances of subsystems in RobotContainer
- set each CanSparksMax instances in Robot to coast when disabled, after determining that we are in a real match

* formatted

* fixes: remove shooter rollers and conveyor, switch to brake in auto and teleop init

* fixed more comments

Panda8188 committed:
Change in shooter angles while moving util

* Calculate change in "azimuth" angle based on desired velocity
* Calculate change in "elevation" angle based on desired velocity

For shooter
*TODO: Implement util in shooter code given current tangential, radial velocity, distance from speaker, and current azimuth angle (needed unfortunately). simple tan logic.

balghane commented on src/main/java/frc/robot/utils/DeltaAngleSpeedCalcUtil.java:
```diff
@@ -0,0 +1,46 @@
+     * https://docs.google.com/document/d/1rIMmzDA1EMH7NQexIMrNYAD-_KFLe9cL4zPLXdl0lek/edit?usp=sharing
+     */
+    double calculatedAzimuthAngleDegrees = Math.atan(tangentialVelocityMPS / SHOOTER_SPEED_MPS);
+    double deltaAzimuthAngleDegrees = currentAzimuthAngleDegrees - calculatedAzimuthAngleDegrees;
```
Isn't `calculatedAzimuthAngleDegrees` itself the delta?

Panda8188 replied:
Actually I think you're right.. good point.

balghane commented on src/main/java/frc/robot/utils/DeltaAngleSpeedCalcUtil.java:
```diff
@@ -0,0 +1,46 @@
+
+public class DeltaAngleSpeedCalcUtil {
+  private static final double SPEAKER_HEIGHT_INCHES = Units.inchesToMeters(77.9528);
+  private static final double SHOOTER_SPEED_MPS = Constants.PLACEHOLDER_DOUBLE;
```
We're going to have to measure this at some point. For now can you set it to like 10 or something?

Also it won't be possible to unit test without setting this to some non-zero value.

Can you make sure your unit test still passes if we change this number? Like maybe make this public and non-final, and set it in your unit test. Or make this variable and the below function non-static, and set this variable in the constructor. You've got different options.

Panda8188 replied:
Alright

voidSevenSevenSix committed:
Trap lineup (#29)

* Code to pull up to and then line up with the trap

* minor fix

* trap lineup

* Basel Fixes I

* Basel Fixes II

* added trap offset, fixed command to actually work

* Basel Fixes III

* Formatting fixes

---------

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

tails618 committed:
Update WPILib to 2024.3.1 (#39)

avakimc commented on src/main/java/frc/robot/utils/DeltaAngleSpeedCalcUtil.java:
```diff
@@ -0,0 +1,46 @@
+import frc.robot.Constants;
+
+public class DeltaAngleSpeedCalcUtil {
+  private static final double SPEAKER_HEIGHT_INCHES = Units.inchesToMeters(77.9528);
```
unless I'm misinterpreting, wouldn't inchesToMeters() return a value in meters? so shouldn't SPEAKER_HEIGHT_INCHES actually be SPEAKER_HEIGHT_METERS?

Panda8188 replied:
Great point thanks 
I will be making the google doc much more easier to read but here's the link https://docs.google.com/document/d/1rIMmzDA1EMH7NQexIMrNYAD-_KFLe9cL4zPLXdl0lek/edit?usp=sharing
its kinda long so sorry about that

the issue is the diagrams kinda take a little to make because i have to replicate it from my own notes and sometimes my notes are illegible..

avakimc reviewed (commented):
small comment. i am not approving because I don't really understand the math

balghane committed:
Practice Field Changes 02/18-22 (#44)

* Cleaned up changes

* Formatting fixes

---------

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

balghane committed:
auto-set-x when not moving

balghane committed:
megatag

tails618 committed:
A bunch of changes

tails618 committed:
More changes

tails618 committed:
✨Changes galore✨

tails618 committed:
So many changes! 🤯

tails618 committed:
set robot position in drive subsystem during disabledPeriodic

tails618 committed:
drive to trap (and lots of other changes!) 🧑‍💻🧠🥳

avakimc committed:
switched pipeline in climb prep

fixed some issues with adding intake limelight

balghane committed:
Crush note on prep climb

balghane committed:
Added Safe Deploy and Don't Move Upon Reenabling

Panda8188 committed:
fixed a bunch of things

* Updated variable names
* Made UnitTest given arbitrary values
* ASK: Should it be calculated - current or current - calculated? The difference in these is the sign.
Will keep working on google doc as a result of these changes

Panda8188 committed:
more tests

removed useless import
MORE TESTS
* 'random" velocity signs (pos/neg)
* "random" speed of note (any value within reason)
* "random" distances from speaker

Panda8188 committed:
spelling fix

Panda8188 committed:
Update DeltaAngleSpeedCalcUtil.java

don't need to set speed initially since its already being set in the constructor

tails618 committed:
Practice field changes 🤖

tails618 committed:
Changes! Wow!

tails618 committed:
biiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiig changes

Panda8188 committed:
implemented feeding to score

* implemented feeding to score so we take note and shoot it over the stage near the speaker and amp to prep for our alliance or ourselves
* commented out a() for temporary usage of a() as feedprepscore
new enum for shooter

balghane committed:
Changes at lab

balghane committed:
read raw values from elevator encoders

balghane committed:
adjust amp score go home logic

avakimc committed:
little commit

avakimc committed:
some changes

tails618 committed:
Control changes

voidSevenSevenSix committed:
removed bindings

voidSevenSevenSix committed:
fixed missing parenthesis

voidSevenSevenSix committed:
Added unclimb command

tails618 committed:
Changes

Panda8188 committed:
first try implementation of deltaAngleSpeedCalcUtil

Panda8188 committed:
completed implementation of deltaanglespeedcalcutil into speakerprepscore.. hopefully

Panda8188 committed:
removed parallel command group

Panda8188 committed:
updated speed of note to be a tad more accurate based on mr. fuller's findings :pray_emoji:

Panda8188 committed:
wow we found a bug woopdeedoo... scratch we found two for twogit add . - thanks ruben

Panda8188 committed:
fixed unit tests

Panda8188 committed:
fixed a couple things, mainly calculating tangential & radial velocity given chassis speeds, and using the current angle and adding it to the delta to actually use the new angle lmao

Panda8188 committed:
similar change from last commit

Panda8188 committed:
removed uselesss imports

Panda8188 commented:
ok so everything kinda died here so im going to fix it all by rebasing off of main which is wisco-cleanup.
